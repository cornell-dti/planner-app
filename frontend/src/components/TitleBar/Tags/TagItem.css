.ColorConfigItem {
  position: relative;
  width: 100%;
  margin: 0;
  padding: 16px 0;
  border-bottom: 2px solid white;
}

.TagName {
  display: inline-block;
  font-weight: bold;
}

.ClassExpandedTitle {
  display: inline-block;
  width: calc(100% - 150px);
}

.DeleteTag {
  width: 50px;
  margin: 0;
  padding: 0;
  color: white;
}

.AddTagName {
  width: calc(100% - 150px);
  background: none;
  border: none;
  color: white;
}

.AddTagName::placeholder {
  color: rgb(188, 188, 188);
}
