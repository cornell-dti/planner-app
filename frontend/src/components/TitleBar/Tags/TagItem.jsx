// @flow strict

import React from 'react';
import type { Node } from 'react';
import { Icon } from 'semantic-ui-react';
import type { EditTagAction, RemoveTagAction } from '../../../store/action-types';
import { editTag as editTagAction, removeTag as removeTagAction } from '../../../store/actions';
import type { Tag } from '../../../store/store-types';
import styles from './TagItem.css';
import { dispatchConnect } from '../../../store/react-redux-util';
import ColorEditor from './ColorEditor';
import { disableBackend } from '../../../util/config';

type Props = {|
  +tag: Tag;
  +editTag: (tag: Tag) => EditTagAction;
  +removeTag: (tagId: number) => RemoveTagAction
|};

/**
 * The tag item component.
 */
function TagItem({ tag, editTag, removeTag }: Props): Node {
  const canBeEdited = disableBackend || tag.id >= 0;
  const onRemove = () => {
    // eslint-disable-next-line
    if (canBeEdited && confirm('Do you want to remove this tag?')) {
      removeTag(tag.id);
    }
  };
  const editColor = (color: string) => {
    if (canBeEdited) {
      editTag({ ...tag, color });
    }
  };
  const editName = (name: SyntheticInputEvent<HTMLInputElement>) => {
    if (canBeEdited) {
      editTag({ ...tag, name: name.currentTarget.value });
    }
  };
  const { name, color, classId } = tag;
  const isClass = classId !== null;
  const nameSplit = name.split(':');
  const nameNode = isClass
    ? <span style={{ width: 'calc(100% - 150px)', display: 'inline-block' }}><span className={styles.TagName}>{nameSplit[0]}</span><span className={styles.ClassExpandedTitle}>{nameSplit[1].trim()}</span></span>
    : <span className={styles.TagName} style={{ width: 'calc(100% - 150px)' }}><input type="text" value={name} onChange={editName} className={styles.TagEdit} /></span>;
  return (
    <li className={styles.ColorConfigItem}>
      {nameNode}
      <ColorEditor color={color} onChange={editColor} />
      <button type="button" className={styles.DeleteTag} onClick={onRemove}>
        <Icon name="close" />
      </button>
    </li>
  );
}

const actionCreators = { editTag: editTagAction, removeTag: removeTagAction };
const Connected = dispatchConnect<Props, typeof actionCreators>(actionCreators)(TagItem);
export default Connected;
