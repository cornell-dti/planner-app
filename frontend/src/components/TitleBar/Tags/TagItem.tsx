import React, { SyntheticEvent, ReactElement } from 'react';
import { Tag } from '../../../store/store-types';
import styles from './TagItem.module.css';
import ColorEditor from './ColorEditor';
import { editTag, removeTag } from '../../../firebase/actions';
import SamwiseIcon from '../../UI/SamwiseIcon';

type Props = { readonly tag: Tag };

/**
 * The tag item component.
 */
export default function TagItem({ tag }: Props): ReactElement {
  const onRemove = (): void => {
    // eslint-disable-next-line no-restricted-globals, no-alert
    if (confirm('Do you want to remove this tag?')) {
      removeTag(tag.id);
    }
  };
  const editColor = (color: string): void => {
    editTag({ ...tag, color });
  };
  const editName = (name: SyntheticEvent<HTMLInputElement>): void => {
    editTag({ ...tag, name: name.currentTarget.value });
  };
  const { name, color, classId } = tag;
  const isClass = classId !== null;
  const nameSplit = name.split(':');
  const nameNode = isClass
    ? (
      <span style={{ width: 'calc(100% - 175px)', display: 'inline-block' }}>
        <span className={styles.TagName}>{nameSplit[0]}</span>
        <span className={styles.ClassExpandedTitle}>{nameSplit[1].trim()}</span>
      </span>
    ) : (
      <span className={styles.TagName} style={{ width: 'calc(100% - 175px)' }}>
        <input type="text" value={name} onChange={editName} className={styles.TagEdit} />
      </span>
    );
  return (
    <li className={styles.ColorConfigItem}>
      {nameNode}
      <ColorEditor color={color} onChange={editColor} />
      <button type="button" className={styles.DeleteTag} onClick={onRemove} tabIndex={0}>
        <SamwiseIcon iconName="x-light" />
      </button>
    </li>
  );
}
