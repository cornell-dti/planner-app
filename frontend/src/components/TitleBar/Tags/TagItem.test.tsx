import React from 'react';
import renderer from 'react-test-renderer';
import TagItem from './TagItem';

it('TagItem without class matches snapshot.', () => {
  const tag = { id: 'foo', order: 1, name: 'Foo', color: 'red', classId: null };
  const tree = renderer.create(<TagItem tag={tag} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('TagItem with class matches snapshot.', () => {
  const tag = { id: 'bar', order: 1, name: 'CS2112: CRITTER', color: 'blue', classId: 'CS2112' };
  const tree = renderer.create(<TagItem tag={tag} />).toJSON();
  expect(tree).toMatchSnapshot();
});
