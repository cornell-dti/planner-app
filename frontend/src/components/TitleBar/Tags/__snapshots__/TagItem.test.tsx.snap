// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagItem with class matches snapshot. 1`] = `
<li
  className="ColorConfigItem"
>
  <span
    style={
      Object {
        "display": "inline-block",
        "width": "calc(100% - 175px)",
      }
    }
  >
    <span
      className="TagName"
    >
      CS2112
    </span>
    <span
      className="ClassExpandedTitle"
    >
      CRITTER
    </span>
  </span>
  <button
    className="ColorEdit"
    onClick={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="ColorDisplay"
      style={
        Object {
          "backgroundColor": "blue",
        }
      }
    />
    <span
      title="More"
    >
      <svg
        alt="More"
        className="Arrow"
        height="1em"
        onKeyUp={[Function]}
        tabIndex={0}
        title="More"
        width="1em"
      >
        v.svg
      </svg>
    </span>
  </button>
  <button
    className="DeleteTag"
    onClick={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      title="Delete"
    >
      <svg
        alt="Delete"
        height="1em"
        onKeyUp={[Function]}
        tabIndex={0}
        title="Delete"
        width="1em"
      >
        XLight.svg
      </svg>
    </span>
  </button>
</li>
`;

exports[`TagItem without class matches snapshot. 1`] = `
<li
  className="ColorConfigItem"
>
  <span
    className="TagName"
    style={
      Object {
        "width": "calc(100% - 175px)",
      }
    }
  >
    <input
      className="TagEdit"
      onChange={[Function]}
      type="text"
      value="Foo"
    />
  </span>
  <button
    className="ColorEdit"
    onClick={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="ColorDisplay"
      style={
        Object {
          "backgroundColor": "red",
        }
      }
    />
    <span
      title="More"
    >
      <svg
        alt="More"
        className="Arrow"
        height="1em"
        onKeyUp={[Function]}
        tabIndex={0}
        title="More"
        width="1em"
      >
        v.svg
      </svg>
    </span>
  </button>
  <button
    className="DeleteTag"
    onClick={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      title="Delete"
    >
      <svg
        alt="Delete"
        height="1em"
        onKeyUp={[Function]}
        tabIndex={0}
        title="Delete"
        width="1em"
      >
        XLight.svg
      </svg>
    </span>
  </button>
</li>
`;
